### PR TITLE
DOC: integrate.solve_ivp: clarify impact of parameter `vectorized`

### LINKED ISSUE #8922: Bug in Solve_ivp with BDF and Radau solvers and numpy arrays
It seems Scipy integration module **solve_ivp** when used with **BDF** and **Radau** solvers appends rounded numbers to the array. This will then cause an error of inconsistent dimensional further on in my code.
I wrote a simple example of a system with (2,1)-dimensional state (i.e. a second order system), which prints the state _y_ as it runs in the function _f(t,y)_, at one point, it appends the rounded number and gave me

```
[[ 13.67999979  13.67999999]
 [-15.99       -15.98999976]]
```

The second column is unnecessary and these added numbers will cause error in my original code.

### Reproducing code example:
```
from numpy import array, matmul
from scipy.integrate import solve_ivp
from numpy import size
def sys(t,x,A,xeq,matmul,size):
    if size(x)==4:
        print(x)
    xdot=matmul(A,x-xeq)
    return xdot

T=10
A=array([[-1,2],[-2,-6]])
x0=array([13.67999999,-15.989999999])
xeq=array([[0.99999999],[2.999999999]])
fun=lambda t, x: sys(t,x,A,xeq,matmul,size)
sol=solve_ivp(fun,[0,T],x0,method='BDF',vectorized=True)
```

### The output is:
```
[[ 13.67999979  13.67999999]
 [-15.99       -15.98999976]]
```

### Scipy/Numpy/Python version information:
```
1.1.0 1.14.4 sys.version_info(major=3, minor=6, micro=5, releaselevel='final', serial=0)
```

### LINKED ISSUE #9061: Will a vectorized fun offer advantages for scipy.integrate.LSODA?
I'm confused with the current [document](https://docs.scipy.org/doc/scipy/reference/generated/scipy.integrate.LSODA.html#scipy.integrate.LSODA), which seems a bit contradictory.

> The vectorized implementation allows a faster approximation of the Jacobian by finite differences (required for this solver).

> A vectorized implementation offers no advantages for this solver. Default is False.

I think we can make that clearer?

### LINKED ISSUE #9728: DOC: scipy.integrate.solve_ivp
https://docs.scipy.org/doc/scipy/reference/generated/scipy.integrate.solve_ivp.html

This omits a crucial piece of information: does vectorized=True correspond to the first option for fun, or the second?

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
Closes gh-8922
Closes gh-9061
Closes gh-9728

#### What does this implement/fix?
gh-9728 noted that the meaning of `vectorized` was not as explicit as it could be.
gh-9061 noted an apparent contradiction in the documentation of `scipy.integrate.LSODA`:

> The vectorized implementation allows a faster approximation of the Jacobian by finite differences (required for this solver).

> A vectorized implementation offers no advantages for this solver. Default is False.

I dug a bit deeper, and found that a vectorized `fun` is [only used by BDF and Radau](https://github.com/scipy/scipy/search?q=fun_vectorized), so it can only be advantageous for these solvers.

However, the [documentation](https://docs.scipy.org/doc/scipy/reference/generated/scipy.integrate.solve_ivp.html) currently guarantees that if `vectorized is True`, `fun` will only be called with 2D `y`. Most of the time, the code wants to call `fun` with 1D `y`, though, so `fun` gets wrapped to accept 1D input and return 1D output.

https://github.com/scipy/scipy/blob/8a64c938ddf1ae4c02a08d2c5e38daeb8d061d38/scipy/integrate/_ivp/base.py#L123-L139

The overhead of using this wrapped function is substantial. It slows down solvers other than BDF and Radau considerably, and it can even slow down BDF and Radau overall unless the bottleneck is in the Jacobian calculation.

```python3
from scipy import integrate
def lotkavolterra(t, z):
    a, b, c, d = (1.5, 1, 3, 1)
    x, y = z
    return [a*x - b*x*y, -c*y + d*x*y]

%timeit integrate.solve_ivp(lotkavolterra, [0, 15], [10, 5], method='Radau')
37.6 ms ± 1.49 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

%timeit integrate.solve_ivp(lotkavolterra, [0, 15], [10, 5], method='Radau', vectorized=True)
44.7 ms ± 200 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

This PR rephrases the documentation to:
- be more explicit (e.g. define `n`, address gh-9728)
- clarify when using `vectorized` is beneficial and when it isn't
- suggest that if `vectorized` is True, `fun` must accept _either_ 1D or 2D `y`. If we can rely on this behavior, we can avoid wrapping `fun` and get the best of both worlds. In many cases, this requires no special effort from the user. For instance, `lotkavolterra` above works exactly how we'd want it to for ND `y`.

#### Additional information
What else should we do?

- Raise a warning when `vectorized=True` but `method not in {'Radau', 'BDF'}`
- Check whether `fun` does the right thing with 1D or 2D `y`, wrap it only if it doesn't, and possibly raise a `DeprecationWarning` if it doesn't.

### CODE REVIEW

**Comment by Patol75** [comment]
Thanks for looking into this @mdhaber. I also remember as a first-time user of `solve_ivp` being extremely confused by the meaning of `vectorized`. Your changes add clarity to the purpose of this argument and, in my opinion, are much welcomed!

About the warnings you mention in **Additional information**, were you thinking to have them implemented as part of this PR or another one?

**Comment by mdhaber** [comment]
> were you thinking to have them implemented as part of this PR or another one?

Let's make them follow-up PRs. It will be easier to review this PR as a doc-only change.

In any case, which of those other things do you think we should do?

**Comment by Patol75** [comment]
I would tend to say that raising a warning if a method other than `Radau` or `BDF` is used with `vectorized=True` should be sufficient given the updated documentation. I am happy to approve the PR if you want it merged relatively soon.

**Comment by mdhaber** [comment]
Thanks @Patol75. Yes, if you wouldn't mind reviewing this one, it would encourage a maintainer to take a look. After this is merged, I'll open a PR to raise warnings.

File: scipy/integrate/_ivp/base.py
Comment (Patol75) [inline scipy/integrate/_ivp/base.py:88]: Instead of `result`, perhaps `returned array`?

File: scipy/integrate/_ivp/base.py
Comment (Patol75) [inline scipy/integrate/_ivp/base.py:92]: Quotes are inconsistent around `Radau`.

File: scipy/integrate/_ivp/base.py
Comment (Patol75) [inline scipy/integrate/_ivp/base.py:93]: What about the phrasing: "and can also negatively impact"

File: scipy/integrate/_ivp/bdf.py
Comment (Patol75) [inline scipy/integrate/_ivp/bdf.py:91]: Should not this part be updated as well?

**Review by Patol75 — CHANGES_REQUESTED** [review]
Thanks again for looking into this @mdhaber. I have left a few minor comments for you to look at.

File: scipy/integrate/_ivp/base.py
Comment (mdhaber) [reply scipy/integrate/_ivp/base.py:92]:
```suggestion
        approximation of the Jacobian by methods 'Radau' and 'BDF', but
```

File: scipy/integrate/_ivp/base.py
Comment (mdhaber) [inline scipy/integrate/_ivp/base.py:89]:
```suggestion
        the returned array is the time derivative of the state corresponding
        with a column of ``y``).
```

File: scipy/integrate/_ivp/base.py
Comment (mdhaber) [reply scipy/integrate/_ivp/base.py:93]:
```suggestion
        will result in slower execution for other methods
        and can also negatively impact 'Radau' and
```

File: scipy/integrate/_ivp/bdf.py
Comment (mdhaber) [reply scipy/integrate/_ivp/bdf.py:91]: Oops!

File: scipy/integrate/_ivp/base.py
Comment (mdhaber) [reply scipy/integrate/_ivp/base.py:93]: How about I split the sentence?

>      will result in slower execution for other methods. It can also
>      result in slower overall execution for 'Radau' and 'BDF' in some
>      circumstances (e.g. small ``len(y0)``).

**Comment by mdhaber** [comment]
Thanks @Patol75. Hopefully addressed everything in [8b9d6ba](https://github.com/scipy/scipy/pull/17836/commits/8b9d6ba190646433c6d9153deba972898739da76).

File: scipy/integrate/_ivp/base.py
Comment (Patol75) [reply scipy/integrate/_ivp/base.py]: Yes, that reads well.

**Review by Patol75 — APPROVED** [review]

**Review by tupui — APPROVED** [review]
LGTM as well, thanks Matt and @Patol75 for the review.

I agree about making a followup to raise a warning when changing the param does nothing.

**Pull request merged**
